### PR TITLE
Disable automatic return value generation

### PR DESCRIPTION
**Pull request by lstrojny**
The automatic return value inference makes it oftentimes harder to do wider refactorings across bigger code bases. This PR should serve as a discussion starter to add a mocking mode, that turns this feature off.

## TODO
- [ ] What to do with `createMock()`, `createConfiguredMock()` etc. Idea would be to add `createStrictMock()`, `createConfiguredStrictMock()`

### CODE REVIEW

**Comment by sebastianbergmann** [comment]
Can you elaborate on the problem you are trying to solve? As far as I understand it right now, this only creates new problems (due to `TypeError`s when stubbed methods that have a return type declaration are called but have no configured return value).

**Comment by lstrojny** [comment]
So right now once I create a mock but not set up expectations for a method, it tries to come up with a default (in `PHPUnit\Framework\MockObject\Invocation::generateReturnValue()`). Let’s imagine this diff in a SUT:

```php
class Collaborator
{
    public function execute(Something $something)
    {
 -          if ($something->isSomething()) {
 +         if (!$something->isSomethingElse()) {
               // some additional behavior
          }
    }
}
```

My testcase will now silently test the opposite case and it will be quite hard to understand from the failing test what the cause is. I would much rather have the test telling me: "you forgot to set up an expectation for a isSomethingElse()". That way I know what the change is and how to change the test accordingly.

**Comment by sebastianbergmann** [comment]
Even if "automatic return value inference" is made optional it will 1) remain enabled by default and 2) no new convenience methods (`createStrictMock()`, ...) will be added. This option is the perfect use case for the Mock Builder API.

**Comment by lstrojny** [comment]
I agree that the automatic generation must stay the default but after seeing it play out in real world projects I don’t think it’s a good default (just like before `createMock()` the defaults for `getMock()` weren't good). I think test suites would greatly benefit from more explicit dependency setup so I think it would be worth thinking about providing convenience methods along the lines of `createMock()` and `createConfiguredMock()`.

**Comment by lstrojny** [comment]
That being said, we can certainly start with making the behavior accessible through the mock builder only first and see where it goes. Either it becomes more popular and we want to add shortcuts later or it doesn’t.

Do you have any feedback on the naming or the route the implementation goes down?

File: src/Invocation/StaticInvocation.php [hunk 1]
```diff
@@ -135,6 +135,10 @@ public function isReturnTypeNullable(): bool
      */
     public function generateReturnValue()
     {
+        if (\strtolower($this->getMethodName()) === '__tostring') {
+            return '';
+        }
+
```
Comment (sebastianbergmann) [inline src/Invocation/StaticInvocation.php]: Can you please send a separate pull request just for this change? Thanks!

**Comment by sebastianbergmann** [comment]
I need to think some more about the naming.

File: src/Invocation/StaticInvocation.php (hunk 1 above)
Comment (lstrojny) [reply src/Invocation/StaticInvocation.php]: Sure thing

Comment (lstrojny) [reply src/Invocation/StaticInvocation.php]: Done, see #407

**Comment by lstrojny** [comment]
On naming, these are my two ideas right now:
 - return value inference
 - return value generation

**Comment by sebastianbergmann** [comment]
"return value generation" makes more sense to me.

**Comment by lstrojny** [comment]
Adjusted naming, added method to the MockObject interface and added a (hacky) way to handle `__toString()`, where we cannot throw an exception.

File: src/InvocationMocker.php [hunk 2]
```diff
@@ -148,6 +158,54 @@ public function invoke(Invocation $invocation)
...
+                sprintf(
+                    "Return value inference disabled and no expectation set up for %s::%s()",
+                    $invocation->getClassName(),
+                    $invocation->getMethodName()
+                )
+            );
+
+            if (\strtolower($invocation->getMethodName()) === '__tostring') {
+                $this->addMatcher(
+                    new class($exception) extends StatelessInvocation
```
Comment (sebastianbergmann) [inline src/InvocationMocker.php]: Does this have to be an anonymous class?

Comment (lstrojny) [reply src/InvocationMocker.php]: Absolutely not. My line of thinking was that I don't see any kind of possibly reuse for this workaround so I made it an inner class. I could extract a proper class and call it something like "DeferredError" or so.

**Comment by lstrojny** [comment]
Updates:
 - Extracted inner class to `DeferredError`
 - Updated all tests
 - Sprinkled a few type hints

**Comment by lstrojny** [comment]
@sebastianbergmann ready to merge from my POV

**Comment by sebastianbergmann** [comment]
Thanks!
